### DISCUSSION TITLE
What to do if a peer dependency is a regular dependency of a subdependency?

### DISCUSSION THREAD

**Discussion by zkochan** [pnpm/pnpm#3963] [General]
Let's assume we want our package/project to have react as a peer dependency.

```json
{
  "name": "card",
  "dependencies": {
    "button": "1.0.0"
  },
  "devDependencies": {
    "react": "16"
  },
  "peerDependencies": {
    "react": "16"
  }
}
```

However, one of the dependencies of our project has react as a prod dependency (and it doesn't match the version that we want). This is the `package.json` of button:

```json
{
  "name": "button",
  "version": "1.0.0",
  "dependencies": {
     "react": "15"
  }
}
```

Should pnpm print a warning in this case?
Should we forcefully convert any prod dependency to a peer dependency if the parent package has the given dependency in peers?
Should we support some new field in `peerDependenciesMeta`? Like `force` for instance?

**Comment by larixer** [comment]
First, it would be great to clarify from the npm team the semantics of peer dependencies in their docs:
https://docs.npmjs.com/cli/v7/configuring-npm/package-json#peerdependencies

Second, if the way I understand their docs is right, I think `pnpm` should not emit a warning in this case, because the resulting dependency graph will look like this:
```
.
├── card
│   └── button@1.0.0
│       └── react@15
└── react@16
```
and the `card` is installed **_along_** with `react@16` peer dependency it has declared. From the `npm` docs the concept of peer dependencies revolves around **_along_**, **_plugin_** and **_host_** words: the `card` _plugin_ should be installed _along_ the `react@16` _host_ and this requirement is fulfilled by the dependency graph above.
